doc/adrs/161_add_support_for_actions_in_tree_item_menu.adoc

```diff
@@ -0,0 +1,180 @@
+
+=== GraphQL
+
+First we need to define a new query that returns the list of tree item contextual menu action.
+
+```graphql
+type TreeDescription implements RepresentationDescription {
+  ...
+  contextMenu(treeItemId: ID!): [TreeItemContextActionMetadata!]!
+  ...
+}
+
+interface TreeItemContextActionMetadata {
+  id: ID!
+  label: String!
+  iconURL: [String!]!
+}
+
+type TreeItemSimpleContextActionMetadata implements TreeItemContextActionMetadata {
+  id: ID!
+  label: String!
+  iconURL: [String!]!
+}
+
+type TreeItemFetchContextActionMetadata implements TreeItemContextActionMetadata {
+  id: ID!
+  label: String!
+  iconURL: [String!]!
+  urlToFetch : String!
+  fetchKind : FetchKind!
+}
+
+enum FetchKind {
+  DOWNLOAD
+  OPEN
+}
+```
+
+For executing an action, we need to add a mutation to call the selected action when it is a simple action. 
+
+```graphql
+extend type Mutation {
+  ...
+  invokeTreeItemContextAction(input: InvokeTreeItemSimpleContextActionInput!): InvokeTreeItemSimpleContextActionPayload!
+  ...
+}
+
+type InvokeTreeItemSimpleContextActionInput {
+  id: ID!
+  editingContextId: ID!
+  representationId: ID!
+  treeItemId: ID!
+  actionId: ID!
+}
+```
+
+For fetch actions, the front end has already the URL to fetch and the kind of fetch to perform inside metadata returned by the back end.
+
+=== Front end
+
+==== Menu 
+
+When the contextual menu of a tree item is open, the front end should request the menu entries to back end.
+The input of the query should contain both the tree id and the tree item id.
+The response should contain a sorted list of menu action metadata.
+This list is used to build the contextual menu of the current tree item.
+
+```typescript
+export interface GQLTreeItemContextActionMetadata {
+  id: string;
+  label: string;
+  iconURL: string[];
+  __typename: string;
+}
+
+export interface GQLTreeItemFetchContextActionMetadata extends GQLTreeItemContextActionMetadata {
+  urlToFetch: string;
+  fetchKind: GQLTreeItemFetchKind;
+}
+
+export type GQLTreeItemFetchKind = 'DOWNLOAD' | 'OPEN';
+```
+
+The `<TreeItemContextMenu>` react component should be updated to add menu entries using the response of the `contextMenu` query.
+
+==== Launching an action
+
+When an action is selected, according to its typename (`__typename`) if it is a simple action, a generic action mutation (`invokeTreeItemContextAction`) is send to the back end to perform the chosen action.
+The input of the mutation should contain:
+
+* the action id
+* the editing context id
+* the tree id
+* the tree item id
+
+As far as fetch actions are concerned, front end has already the URL to fetch and knows the way the fetch should be performed, so no more exchange with backend is needed.
+
+=== Back end
+
+The TreeDescription is updated to have a new property : 
+
+```java
+public final class TreeDescription implements IRepresentationDescription {
+  ...
+  List<TreeItemContextAction> contextActions;
```

Comment:
This has been done in the diagram part and it was a terrible decision. It has been horrible to move back from this and we are still not fully done yet. Just ask for a list of `ITreeItemContextMenuEntryProvider` in the event handler used to handle the input and create one to retrieve the view model to figure out the menu entry from there.

I don't want to relive the experience of what we went through with `DiagramDescription#palettes`, it's way too static, it's such a mess that we are still not done rewriting the whole thing. see `org.eclipse.sirius.components.collaborative.diagrams.dto.ITool` vs `org.eclipse.sirius.components.diagrams.tools.ITool` for example and the whole code using one or the other...

Reply:
Understood, this will be done in a second time.
I mention this second phase in the adr